File: openlogs-sdk-csharp/src/OpenLogs.Sdk/OpenLogsChainService.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace OpenLogs.Sdk;

public sealed class OpenLogsChainService
{
    public const string Spec = "openlogs.v2";

    private readonly List<OpenLogsRecord> _records = [];
    private string? _latestHash;

    public OpenLogsRecord Log(OpenLogsEntry entry)
    {
        var id = Guid.NewGuid().ToString();
        var prev = _latestHash;
        var hash = OpenLogsSigner.ComputeHash(Spec, id, entry, prev);

        var record = new OpenLogsRecord(Spec, id, entry, prev, hash, null);
        _records.Add(record);
        _latestHash = record.Hash;
        return record;
    }

    public OpenLogsRecord LogSigned(OpenLogsEntry entry, byte[] privateKey, byte[] publicKey, string? kid = null)
    {
        var unsigned = Log(entry);
        var signed = OpenLogsSigner.SignRecord(unsigned, privateKey, publicKey, kid);

        _records[^1] = signed;
        _latestHash = signed.Hash;
        return signed;
    }

    public IReadOnlyList<OpenLogsRecord> GetChain() => _records.ToList();

    public string? GetLatestHash() => _latestHash;

    public bool VerifyChain()
    {
        string? expectedPrev = null;
        foreach (var record in _records)
        {
            if (record.Prev != expectedPrev)
            {
                return false;
            }

            var recomputed = OpenLogsSigner.ComputeHash(record.Spec, record.Id, record.Entry, record.Prev);
            if (recomputed != record.Hash)
            {
                return false;
            }
            expectedPrev = record.Hash;
        }
        return true;
    }

    public bool VerifySignatures(byte[] publicKey)
    {
        foreach (var record in _records)
        {
            if (record.Signature is null)
            {
                continue;
            }
            if (!OpenLogsSigner.VerifyRecordSignature(record, publicKey))
            {
                return false;
            }
        }
        return true;
    }
}

public static class OpenLogsSigner
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = false,
    };

    public static (byte[] PrivateKey, byte[] PublicKey) GenerateEd25519Keypair()
    {
        var privateKey = new byte[32];
        RandomNumberGenerator.Fill(privateKey);
        var publicKey = new byte[32];
        Ed25519.PublicKeyFromSeed(privateKey, publicKey);
        return (privateKey, publicKey);
    }

    public static string ComputeHash(string spec, string id, OpenLogsEntry entry, string? prev)
    {
        var canonical = SigningPayload(spec, id, entry, prev, null);
        var digest = SHA256.HashData(Encoding.UTF8.GetBytes(canonical));
        return Convert.ToHexString(digest).ToLowerInvariant();
    }

    public static OpenLogsRecord SignRecord(OpenLogsRecord record, byte[] privateKey, byte[] publicKey, string? kid = null)
    {
        var payload = SigningPayload(record.Spec, record.Id, record.Entry, record.Prev, record.Hash);
        var data = Encoding.UTF8.GetBytes(payload);

        var signature = new byte[64];
        Ed25519.Sign(data, privateKey, signature);

        return record with
        {
            Signature = new OpenLogsSignature(
                Alg: "Ed25519",
                PublicKeyHex: Convert.ToHexString(publicKey).ToLowerInvariant(),
                ValueHex: Convert.ToHexString(signature).ToLowerInvariant(),
                Kid: kid
            )
        };
    }

    public static bool VerifyRecordSignature(OpenLogsRecord record, byte[] publicKey)
    {
        if (record.Signature is null)
        {
            return false;
        }

        var payload = SigningPayload(record.Spec, record.Id, record.Entry, record.Prev, record.Hash);
        var data = Encoding.UTF8.GetBytes(payload);
        var sig = Convert.FromHexString(record.Signature.ValueHex);
        return Ed25519.Verify(sig, data, publicKey);
    }

    private static string SigningPayload(string spec, string id, OpenLogsEntry entry, string? prev, string? hash)
    {
        var dataSorted = entry.Data is null
            ? new SortedDictionary<string, object?>()
            : new SortedDictionary<string, object?>(entry.Data);

        var indexesSorted = entry.Indexes is null
            ? new SortedDictionary<string, string>()
            : new SortedDictionary<string, string>(entry.Indexes);

        var payload = new Dictionary<string, object?>
        {
            ["spec"] = spec,
            ["id"] = id,
            ["prev"] = prev,
            ["entry"] = new Dictionary<string, object?>
            {
                ["actor"] = entry.Actor,
                ["tps"] = entry.Tps,
                ["event"] = entry.Event,
                ["data"] = dataSorted,
                ["indexes"] = indexesSorted,
            },
        };

        if (hash is not null)
        {
            payload["hash"] = hash;
        }

        return JsonSerializer.Serialize(payload, JsonOptions);
    }
}

File: openlogs-sdk-java/src/main/java/com/nextera/one/openlogs/sdk/OpenLogsChainService.java
package com.nextera.one.openlogs.sdk;

import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;

public class OpenLogsChainService {
  private static final String SPEC = "openlogs.v2";

  private final List<OpenLogsRecord> records = new ArrayList<>();
  private String latestHash;

  public synchronized OpenLogsRecord log(OpenLogsEntry entry) {
    Objects.requireNonNull(entry, "entry is required");

    String id = UUID.randomUUID().toString();
    String prev = latestHash;

    String hash = computeHash(SPEC, id, entry, prev);
    OpenLogsRecord record = new OpenLogsRecord(SPEC, id, entry, prev, hash);

    records.add(record);
    latestHash = hash;
    return record;
  }

  public synchronized OpenLogsRecord logSigned(
      OpenLogsEntry entry,
      PrivateKey privateKey,
      PublicKey publicKey,
      String kid
  ) {
    OpenLogsRecord unsigned = log(entry);
    OpenLogsRecord signed = OpenLogsSigner.signRecord(unsigned, privateKey, publicKey, kid);

    records.set(records.size() - 1, signed);
    latestHash = signed.getHash();
    return signed;
  }

  public synchronized List<OpenLogsRecord> getChain() {
    return List.copyOf(records);
  }

  public synchronized String getLatestHash() {
    return latestHash;
  }

  public synchronized boolean verifyChain() {
    String expectedPrev = null;
    for (OpenLogsRecord record : records) {
      if (!Objects.equals(record.getPrev(), expectedPrev)) {
        return false;
      }

      String recomputed = computeHash(
          record.getSpec(),
          record.getId(),
          record.getEntry(),
          record.getPrev()
      );

      if (!Objects.equals(recomputed, record.getHash())) {
        return false;
      }

      expectedPrev = record.getHash();
    }
    return true;
  }

  public synchronized boolean verifySignatures(PublicKey publicKey) {
    for (OpenLogsRecord record : records) {
      if (record.getSignature() == null) {
        continue;
      }
      if (!OpenLogsSigner.verifySignature(record, publicKey)) {
        return false;
      }
    }
    return true;
  }

  static String computeHash(String spec, String id, OpenLogsEntry entry, String prev) {
    Map<String, Object> payload = new LinkedHashMap<>();
    payload.put("spec", spec);
    payload.put("id", id);
    payload.put("prev", prev);

    Map<String, Object> entryPayload = new LinkedHashMap<>();
    entryPayload.put("actor", entry.getActor());
    entryPayload.put("tps", entry.getTps());
    entryPayload.put("event", entry.getEvent());
    entryPayload.put("data", entry.getData());
    entryPayload.put("indexes", entry.getIndexes());
    payload.put("entry", entryPayload);

    return Hashing.sha256Hex(CanonicalJson.stringify(payload));
  }
}
